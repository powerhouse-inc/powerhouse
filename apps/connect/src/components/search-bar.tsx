import {
    ConnectSearchBar,
    ConnectSearchBarProps,
    Icon,
} from '@powerhousedao/design-system';
import { useTranslation } from 'react-i18next';

const defaultFilterItems: ConnectSearchBarProps['filterItems'] = [
    {
        id: 'project',
        label: '.project',
        icon: <Icon name="project" color="#FF6A55" size={16} />,
    },
    {
        id: 'budget',
        label: '.budget',
        icon: <Icon name="bar-chart" color="#8E55EA" size={16} />,
    },
    {
        id: 'profile',
        label: '.profile',
        icon: <Icon name="person" color="#3E90F0" size={16} />,
    },
    {
        id: 'legal',
        label: '.legal',
        icon: <Icon name="briefcase" color="#4BAB71" size={16} />,
    },
    {
        id: 'atlas',
        label: '.Atlas',
        icon: <Icon name="globe" color="#FF8A00" size={16} />,
    },
];

export const SearchBar = () => {
    const { t } = useTranslation();

    return (
        <ConnectSearchBar
            className="m-4 max-w-[642px] shrink-0 bg-[#F4F4F4]"
            placeholder={t('searchbar.placeholder')}
            filterLabel={t('searchbar.filterLabel')}
            filterItems={defaultFilterItems}
        />
    );
};
