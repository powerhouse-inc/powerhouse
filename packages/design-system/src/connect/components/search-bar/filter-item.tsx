import { mergeClassNameProps } from "@/powerhouse";

export type FilterItemType = {
  id: string;
  label: string;
  icon?: React.JSX.Element;
};

export interface FilterItemProps extends React.HTMLAttributes<HTMLDivElement> {
  item: FilterItemType;
}

export const FilterItem: React.FC<FilterItemProps> = (props) => {
  const { item, ...containerProps } = props;

  return (
    <div
      {...mergeClassNameProps(
        containerProps,
        "flex h-full flex-row items-center justify-between gap-x-4 px-2",
      )}
    >
      {item.icon}
      <div className="text-sm font-semibold text-slate-200">{item.label}</div>
    </div>
  );
};
