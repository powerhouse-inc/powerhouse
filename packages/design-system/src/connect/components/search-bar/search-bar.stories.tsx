import { Icon } from "@/powerhouse";
import type { Meta, StoryObj } from "@storybook/react";
import { ConnectSearchBar, ConnectSearchBarProps } from "./search-bar";

const filterItems: ConnectSearchBarProps["filterItems"] = [
  {
    id: "project",
    label: ".project",
    icon: <Icon className="text-red-700" name="Project" size={16} />,
  },
  {
    id: "budget",
    label: ".budget",
    icon: <Icon className="text-purple-900" name="BarChart" size={16} />,
  },
  {
    id: "profile",
    label: ".profile",
    icon: <Icon className="text-blue-900" name="Person" size={16} />,
  },
  {
    id: "legal",
    label: ".legal",
    icon: <Icon className="text-green-900" name="Briefcase" size={16} />,
  },
  {
    id: "atlas",
    label: ".Atlas",
    icon: <Icon className="text-orange-900" name="Globe" size={16} />,
  },
];

const meta: Meta<typeof ConnectSearchBar> = {
  title: "Connect/Components/SearchBar",
  component: ConnectSearchBar,
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  args: {
    placeholder: "Search Files",
    filterLabel: "File type",
    filterItems,
  },
};
