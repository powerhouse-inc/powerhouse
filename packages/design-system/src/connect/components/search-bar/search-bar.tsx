import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuItem,
  DropdownMenuTrigger,
  Icon,
} from "@/powerhouse";
import { ChangeEvent, useMemo } from "react";
import { twMerge } from "tailwind-merge";
import { FilterItem, FilterItemType } from "./filter-item";

export interface ConnectSearchBarProps {
  value: string;
  onChange: (value: string) => void;
  placeholder?: string;
  filterLabel?: string;
  filterItems?: Array<FilterItemType>;
  selectedFilter?: string;
  onFilterSelect?: (filterId: string) => void;
}

export const ConnectSearchBar: React.FC<ConnectSearchBarProps> = (props) => {
  const {
    value,
    onChange,
    placeholder,
    filterLabel,
    filterItems,
    selectedFilter,
    onFilterSelect = () => {},
  } = props;

  const items = useMemo(
    () =>
      filterItems?.map((item) => ({
        id: item.id,
        content: <FilterItem item={item} />,
      })) ?? [],
    [filterItems],
  );

  const selectedItemFilter = filterItems?.find(
    (item) => item.id === selectedFilter,
  );

  const filterLabelContent = selectedItemFilter ? (
    <FilterItem className="gap-x-1" item={selectedItemFilter} />
  ) : (
    filterLabel && (
      <div className="mr-2 text-sm font-semibold text-slate-200">
        {filterLabel}
      </div>
    )
  );

  function handleChange(event: ChangeEvent<HTMLInputElement>) {
    onChange(event.target.value);
  }

  return (
    <div className="flex items-center">
      <Icon className="mr-3" name="Search" />
      <input
        className={twMerge(
          "flex h-[52px] min-w-0 flex-1 items-center rounded-xl border border-gray-200 bg-gray-50 px-4 text-sm text-slate-200 outline-none",
        )}
        onChange={handleChange}
        placeholder={placeholder}
        value={value}
      />
      <DropdownMenu>
        <DropdownMenuTrigger className="ml-3 flex h-full flex-row items-center outline-none">
          {filterLabelContent} <Icon name="ChevronDown" />
        </DropdownMenuTrigger>
        <DropdownMenuContent className="rounded-xl border border-gray-100 bg-gray-50 p-2">
          {items.map((item) => (
            <DropdownMenuItem
              className="h-10 cursor-pointer overflow-hidden rounded-lg hover:bg-gray-100"
              id={item.id}
              key={item.id}
              onSelect={() => onFilterSelect(item.id)}
            >
              {item.content}
            </DropdownMenuItem>
          ))}
        </DropdownMenuContent>
      </DropdownMenu>
    </div>
  );
};
