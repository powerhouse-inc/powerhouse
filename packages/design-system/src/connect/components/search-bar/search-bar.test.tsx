import {
  fireEvent,
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import { ConnectSearchBar, ConnectSearchBarProps } from "./search-bar";

const filterItems: ConnectSearchBarProps["filterItems"] = [
  {
    id: "project",
    label: ".project",
  },
  {
    id: "budget",
    label: ".budget",
  },
  {
    id: "profile",
    label: ".profile",
  },
  {
    id: "legal",
    label: ".legal",
  },
  {
    id: "atlas",
    label: ".Atlas",
  },
];

describe("ConnectSearchBar Component", () => {
  it("should match snapshot", async () => {
    const { asFragment } = render(
      <ConnectSearchBar
        filterItems={filterItems}
        filterLabel="File type"
        onChange={() => {}}
        placeholder="Search Files"
        value="test"
      />,
    );

    await waitForElementToBeRemoved(() =>
      screen.getAllByTestId("icon-fallback"),
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should call onChange when input value is changed", () => {
    const onChange = vi.fn();

    render(
      <ConnectSearchBar
        data-testid="search-bar"
        filterItems={filterItems}
        filterLabel="File type"
        onChange={onChange}
        placeholder="Search Files"
        value=""
      />,
    );

    fireEvent.change(screen.getByRole("textbox"), {
      target: { value: "test" },
    });

    expect(onChange).toBeCalledTimes(1);
    expect(onChange).toBeCalledWith("test");
  });

  it("should display filter selected in filter label", () => {
    render(
      <ConnectSearchBar
        data-testid="search-bar"
        filterItems={filterItems}
        filterLabel="File type"
        onChange={() => {}}
        placeholder="Search Files"
        selectedFilter="project"
        value="test"
      />,
    );

    expect(screen.getByText(".project")).toBeInTheDocument();
    expect(screen.queryByText("File type")).not.toBeInTheDocument();
  });
});
