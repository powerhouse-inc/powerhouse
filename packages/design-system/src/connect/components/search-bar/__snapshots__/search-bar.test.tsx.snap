// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConnectSearchBar Component > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex items-center"
  >
    <svg
      class="mr-3"
      fill="currentcolor"
      style="width: 24px; height: 24px;"
      viewBox="0 0 24 24"
    >
      <path
        d="M10.0137 1.95312C5.59537 1.95312 2.01367 5.53513 2.01367 9.95314C2.01367 14.3712 5.59537 17.9532 10.0137 17.9532C11.8504 17.9532 13.5913 17.3222 14.9416 16.2822L19.2949 20.6722C19.6855 21.0622 20.3419 21.0622 20.7325 20.6722C21.123 20.2812 21.123 19.6252 20.7325 19.2342L16.349 14.8741C17.3898 13.5241 18.0137 11.7901 18.0137 9.95314C18.0137 5.53513 14.432 1.95312 10.0137 1.95312ZM10.0137 3.95313C13.3274 3.95313 16.0137 6.63914 16.0137 9.95314C16.0137 13.2672 13.3274 15.9532 10.0137 15.9532C6.69997 15.9532 4.01367 13.2672 4.01367 9.95314C4.01367 6.63914 6.69997 3.95313 10.0137 3.95313Z"
      />
    </svg>
    <input
      class="flex h-[52px] min-w-0 flex-1 items-center rounded-xl border border-gray-200 bg-gray-50 px-4 text-sm text-slate-200 outline-none"
      placeholder="Search Files"
      value="test"
    />
    <button
      aria-expanded="false"
      aria-haspopup="menu"
      class="ml-3 flex h-full flex-row items-center outline-none"
      data-state="closed"
      id="radix-:r0:"
      type="button"
    >
      <div
        class="mr-2 text-sm font-semibold text-slate-200"
      >
        File type
      </div>
       
      <svg
        fill="none"
        stroke="currentcolor"
        style="width: 24px; height: 24px;"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 9L12 15L18 9"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
